test/rational.jl

```diff
@@ -727,3 +727,166 @@ end
     @test rationalize(1.192 + 2.233im) == 149//125 + 2233//1000*im
     @test rationalize(Int8, 1.192 + 2.233im) == 118//99 + 67//30*im
 end
+
+naive_rational_to_bigfloat(x::Rational) = BigFloat(numerator(x))/BigFloat(denominator(x))
+
+function is_faithfully_rounded(;approximate::AbstractFloat, accurate)
+    p = prevfloat(approximate)
+    n = nextfloat(approximate)
+    (approximate == accurate) && return true
+    (approximate < accurate) ? (accurate < n) : (p < accurate)
+end
+
+is_in_range(::Type{U}, v::Integer) where {U<:Integer} = typemin(U) ≤ v ≤ typemax(U)
+is_in_range(::Type{BigInt}, ::Integer) = true
+
+is_typemin(::Type{U}, v::Integer) where {U<:Integer} = v == typemin(U)
+is_typemin(::Type{BigInt}, ::Integer) = false
+
+has_asymmetric_typemin(::Type{<:Unsigned}) = false
+has_asymmetric_typemin(::Type{<:Signed}) = true
+has_asymmetric_typemin(::Type{BigInt}) = false
+
+function is_asymmetric_typemin(::Type{U}, v::Integer) where {U<:Integer}
+    has_asymmetric_typemin(U) & is_typemin(U, v)
+end
+
+with_unsigned_if_any(::Type{T}) where {T<:Signed} = (T, unsigned(T))
+with_unsigned_if_any(::Type{BigInt}) = (BigInt,)
+
+is_arbitrary_precision_float(::Type{<:AbstractFloat}) = false
+is_arbitrary_precision_float(::Type{BigFloat}) = true
```

Comment:
Are these just `!hastypemax(T)`?

Reply:
Not at all, `typemax` is infinity for all floats anyway, and this function tries to distinguish between arbitrary precision FP and other FP types. A more correct way to implement `is_arbitrary_precision_float` would be like so, though:
```julia
function is_arbitrary_precision_float(::Type{F}) where {F<:AbstractFloat}
    applicable(setprecision, F, 1)
end
is_arbitrary_precision_float(::Type{<:Base.IEEEFloat}) = false
is_arbitrary_precision_float(::Type{BigFloat}) = true
```

I'll make this change.

Reply:
Then I guess it's better to just replace this with a `T === BigFloat` check. I don't see us adding more arbitrary precision floating point types to Base.